include/mesh/sides_to_elem_map.h

```diff
@@ -0,0 +1,145 @@
+// The libMesh Finite Element Library.
+// Copyright (C) 2002-2024 Benjamin S. Kirk, John W. Peterson, Roy H. Stogner
+
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+
+#ifndef LIBMESH_SIDES_TO_ELEM_MAP_H
+#define LIBMESH_SIDES_TO_ELEM_MAP_H
+
+// libMesh includes
+#include "libmesh/id_types.h" // dof_id_type
+#include "libmesh/hashword.h" // Utility::hashword()
+
+// C++ includes
+#include <unordered_map>
+#include <vector>
+
+namespace libMesh
+{
+
+// Forward declarations
+class MeshBase;
+class Elem;
+
+namespace MeshTools
+{
+
+/**
+ * This class implements a generalization of the
+ * MeshTools::build_nodes_to_elem_map() function, but rather than
+ * being a standalone function taking a std::unordered_map argument,
+ * we define a class with APIs for both building the underlying
+ * map data structure and querying its contents.  This way, we avoid
+ * issues with hash collisions resulting from the Elem::low_order_key()
+ * function (which could occur with as few as 100k sides when using a
+ * 32-bit dof_id_type) in addition to being able to control exactly
+ * how the user will interact with the class.
+ *
+ * \author John W. Peterson
+ * \date 2024
+ */
+class SidesToElemMap
+{
+public:
+  /**
+   * Default constructor/destructor
+   */
+  SidesToElemMap();
+  ~SidesToElemMap();
+
+  /**
+   * Static build function. Most users will simply call this
+   * function to construct the SidesToElemMap object for their
+   * Mesh.
+   */
+  static SidesToElemMap build(const MeshBase & mesh);
+
+  /**
+   * Return a const reference to the list of Elems connected to "side"
+   * of "elem". The returned list of Elems will also include "elem"
+   * itself. Throws an error if the requested (elem, side) combination
+   * is not found in the map (this should not happen because every
+   * side in the map should be attached to at least one Elem).
+   */
+  const std::vector<const Elem *> &
+  get_connected_elems(const Elem * elem, unsigned int side) const;
```

Comment:
I'm not thrilled about returning a reference from a const function.  That forces us to either make sure that those vectors will always part of our internal data structure (as opposed to any means of finding that set of elements on the fly) or to do nasty things with `mutable` if we change that.

But I certainly don't want to return the vector by value either and force a copy with heap allocation on every call.

Could we perhaps return a pair of begin/end iterators, of a type that for now is just a `typedef` of `vector::const_iterator`?  That'll still be ABI-breaking to change but it'll be easy to keep API compatible.

Reply:
Yeah, it's not that it forces the function to be conceptually non-const now, it's that if we ever change the data structure it'd force the function to do evil mutable or const_cast things later.  Thanks for the change.